### ISSUE TITLE
Possible Bug in Implementation of 'S' String Format Type

### ISSUE THREAD

**Issue by PorcelainMouse** [lebigot/uncertainties#117]
I have been working through a problem using 'pint' to format LaTeX figures *with* the SIunitx package.  Internally, it uses 'uncertainties' to hold uncertain values, and implements the 'Lx' format spec--a SIunitx-specific LaTeX string format spec--by passing 'S' to 'uncertainties' and then putting the result inside \SI{} macro.

Under the current implementation of both packages, this causes LaTeX compile to fail when the uncertainty value is expressed with a decimal point, e.g. 123.45(1.20), which is what uncertainties will produce in that case.

If 'uncertainties' intends to be producing a "concise notation" figure containing both nominal and uncertainty values, then this is a bug, as best I can tell.  At least, according to NIST (https://physics.nist.gov/cgi-bin/cuu/Info/Constants/definitions.html) and SIunitx.  SIunitx accepts this "concise" format as input, so I would be valuable to the community, I think, for 'uncertainties' to support it, if not with 'S', then with some new format type.

I would be happy to try to implement this for you and send a PR, but it is likely be much easier for you to design the solution you like and implement it.  If you would prefer me to have go, please give me some clear instructions and preferences and I'll try.

See https://github.com/hgrecco/pint/issues/814 for some more background.

**Comment by lebigot** [comment]
Can you make explicit what problem you see with the formatting in uncertainties? Giving an example of expected output would also help.

Also, the NIST reference that is quoted does not address the case that you cite (uncertainty digits on both sides of the decimal point), so I can't yet see that uncertainties' representation of uncertain digits is not an accepted one.  

Thanks!

**Comment by PorcelainMouse** [comment]
Sure, no problem.  In the "concise" notation, () cannot contain a decimal point as it is supposed to contain the uncertain least significant digits.  If that includes digits surrounding the decimal, then you write them without the decimal.  The NIST documentation says "the number in parentheses is the numerical value of the standard uncertainty referred to the corresponding last digits of the quoted result".  It could be more clear, I guess, but that doesn't allow for any decimals, strictly speaking.  For example,

`123.45 +/- 1.20 == 123.45(120) != 123.45(1.20)`

I supposed, if you use scientific notation, this problem is pretty rare, as most numbers that come out of real experiments are not uncertain in the leading digit.  However, if you are in my introductory physics course, then uncertainties > 10 % occur often and, if you don't use scientific notation, then you have this case.

I only noticed this when 'pint' tried to pass a string formatted with 'S' to SIunitx, which threw a parse error.  For example, this

`\SI[separate-uncertainty=true]{10.5(1.0)}{\meter}  % wrong`

will not compile.

I tried to think about how to code this and it is not trivial, especially considering the other type specifications that need to be respected.  Somehow, you have to determine the LSD of nominal value, then produce the uncertainty value with the same LSD, then remove the decimal point, if necessary.  I think. Maybe there is a more clever way to do it, but that is the literal, direct approach.

**Comment by PorcelainMouse** [comment]
This is sort of off topic, but I've been thinking about this issue a lot and I'd like to run this by you.  Are you interested in having 'uncertainties' support a new string conversion type that would use the uncertainty value to automatically determine the sig figs and, therefore, an "optimal" string representation.  It seems to me that, if I specify an uncertainty for any value, then that completely determines the significant figures of the nominal values, since uncertain digits cannot be considered significant.

I think a "natural" way to represent uncertain numbers is all MSDs up to and including the first uncertain digit.  By "natural", I mean in the sense that 'f' and 'g' have a default precision based on common sensibilities, but this is even more logical.  (You could argue that the default should be 0 uncertain digits, and I'd be happy with that, too.)

Anyway, I thought this package would be good place for such an implementation, if you think it has any value.

(I started thinking about this because, if I remember correctly, LSD cannot be determined from the underlying bit representation, which is how I first started to conceptualize this problem.  I don't know enough Python internals to say if there is some extra information there that could help, but, in general, I don't know how to determine the LSD of a stored value, without making some assumptions.  Doesn't 'g' do that?  Then I thought the MSD of the error would actually be a good way to determine the LSD of the nominal value.)

**Comment by lebigot** [comment]
Thanks for the details on the point-straddling uncertainty. I agree that nowhere do they mention the decimal point, so I'm guessing that it shouldn't be there. I will write NIST and ask them for confirmation—there is a very small chance that the `SIunitx` package be non-standard, so I wouldn't want _two_ packages to be non-standard.

As for the "natural" representation of numbers in relation to their most significant digits, have you had a look at the [`u` format that I implemented](https://uncertainties-python-package.readthedocs.io/en/latest/user_guide.html#precision-control)? I am wondering whether the `.1u` format doesn't do something very close to what you want.

**Comment by PorcelainMouse** [comment]
Great!  Good thinking, makes sense.

I believe I have seen the concise form used in print, before, without the decimal when it would otherwise be written, but I cannot remember where I saw it.  I would have linked to it if I had a good example.

OMG, yes, that's exactly what I was thinking!  Hah!  In reading your unit tests, I discovered 'u' again, but, did not read closely enough; I thought it was sig figs of the nominal value.  But, no, your implementation is exactly what I was thinking.  Awesome!

**Comment by lebigot** [comment]
I did contact an expert at NIST: their reference publication that lists the "official" values of fundamental constants does use the same 123.4(5.6) convention as `uncertainties`. Reference: Table IX of the latest (soon to be superseded) paper on the Least Square Adjustment of fundamental constants available at http://physics.nist.gov/constants (as of 2020-01-02).

Another argument for this convention is my original thought: 123.4(56) can make people think that it represents 123.4 ± 0.56 (because of the natural "limit" defined by the decimal point), so it is best to remove any question and be explicit by including the decimal point in the uncertainty.

As a consequence, I would suggest that SIunitx should handle this convention from a major reference source.

As for the formatting of uncertainties, setting the number of digits on the nominal value is done as usual with a format like `.6` (thus, the `.2u` format _does_ bring something new and useful, namely to display the _uncertainty_ on 2 digits).

**Comment by PorcelainMouse** [comment]
Thanks for the research.  Okay, I'll try to get SIunitx maintainers to fix it, there.

**Comment by PorcelainMouse** [comment]
Okay, SIunitx says they are not immediately convinced to change their standard, which they prefer to the one NIST and you prefer.  The question is, where to go from here?

I will have to take a step back and recast the problem as an incompatibility, since there seems to be no agreement on the standard that can guide the path forward. The problem I am trying to solve is that I cannot use uncertainties, nor 'pint', to create code for SIunitx.  (Granted, this is a bit of an edge case, but I ran into it the very first time I tried to put all these pieces together.)  The source of the "problem", AFAICT, is that 'uncertainties' does not implement a string format suitable for SIunitx input, and 'pint''s implementation has a bug, but only because it relies on 'uncertainties''s 'S' type string conversion, which is not compatible with SIunitx due to this disagreement about the radix point.  Am I making sense? I realize this is complicated.

I see some immediate options; there must be more, of course.
1. uncertainty implements a new, format conversion type or option that acts exactly like 'S', but doesn't include the radix point in the uncertainty.
1. uncertainty implements a separate conversion type that is compatible with SIunitx's other formatting rules.  (Neither the default python string repr, nor the 'L' type conversion is valid for SIunitx.)
1. uncertainty leaves it all up to 'pint'.  I think I have submitted an acceptable patch that should work, but there are probably new edge cases I should write unit tests for.

Is any of this, or another option, appealing to you?  There is little chance I can contribute to SIunitx, but I'm confident enough with Python to try to contribute to 'uncertainties', if that is helpful.
